src/lib/storage/value_segment.cpp

```diff
@@ -172,6 +172,157 @@ size_t ValueSegment<T>::estimate_memory_usage() const {
+                             [&](const auto& row_id, const auto& search_value) { return !row_id.is_null(); })));
+      } else {
+        non_null_end = static_cast<ChunkOffset>(
+            std::distance(_null_values.value().cbegin(),
+                          std::lower_bound(_null_values.value().cbegin(), _null_values.value().cend(), true)));
+      }
+    }
+  }
+
+  // std::cout << "non_null_end " << non_null_end << std::endl;
+
+  return non_null_end;
+}
+
+template <typename T>
+ChunkOffset ValueSegment<T>::get_first_offset(const AllTypeVariant& search_value,
+                                              const std::shared_ptr<const PosList>& position_filter) const {
+  Assert(_sort_order, "The segment needs to be sorted to calculate the first bound.");
+
+  const auto non_null_begin = get_non_null_begin_offset(position_filter);
+  const auto non_null_end = get_non_null_end_offset(position_filter);
+
+  const auto casted_search_value = type_cast_variant<T>(search_value);
+
+  // TODO(cmfcmf): Quite a bit of duplication.
+  if (_sort_order.value() == OrderByMode::Ascending || _sort_order.value() == OrderByMode::AscendingNullsLast) {
+    if (position_filter != nullptr) {
+      const auto result = std::lower_bound(
+          position_filter->cbegin(), position_filter->cend(), casted_search_value,
+          [&](const auto& row_id, const auto& search) { return _values[row_id.chunk_offset] < search; });
+      if (result == position_filter->cend()) {
+        return INVALID_CHUNK_OFFSET;
+      }
+      return static_cast<ChunkOffset>(std::distance(position_filter->cbegin(), result));
+    } else {
+      const auto begin = _values.cbegin() + non_null_begin;
+      const auto end = _values.cbegin() + non_null_end;
+      const auto result = std::lower_bound(begin, end, casted_search_value);
+      if (result == end) {
+        return INVALID_CHUNK_OFFSET;
+      }
+      return static_cast<ChunkOffset>(std::distance(_values.cbegin(), result));
+    }
+  } else {
+    if (position_filter != nullptr) {
+      const auto result = std::lower_bound(
+          position_filter->cbegin(), position_filter->cend(), casted_search_value,
+          [&](const auto& row_id, const auto& search) { return _values[row_id.chunk_offset] > search; });
+      if (result == position_filter->cend()) {
+        return INVALID_CHUNK_OFFSET;
+      }
+      return static_cast<ChunkOffset>(std::distance(position_filter->cbegin(), result));
+    } else {
+      const auto begin = _values.cbegin() + non_null_begin;
+      const auto end = _values.cbegin() + non_null_end;
+      const auto result = std::lower_bound(begin, end, casted_search_value, std::greater<T>());
+      if (result == end) {
+        return INVALID_CHUNK_OFFSET;
+      }
+      return static_cast<ChunkOffset>(std::distance(_values.cbegin(), result));
+    }
+  }
+}
+
+template <typename T>
+ChunkOffset ValueSegment<T>::get_last_offset(const AllTypeVariant& search_value,
```

Comment:
The only difference with the function `get_first_offset` is the calculation of the `result` value, the rest is duplicated.
See the [mentioned issue](https://github.com/hyrise/hyrise/pull/1394/files#r257999247) for dictionary segments.